BiometricLoginKotlin/app/src/main/java/com/example/biometricloginsample/LoginActivity.kt

```diff
@@ -0,0 +1,199 @@
+        useBiometricsView.setOnClickListener {
+            if (ciphertextWrapper != null) {
+                encryptedServerTokenWrapper = ciphertextWrapper
+                showBiometricPromptForDecryption()
+            } else {
+                startActivity(Intent(this, EnableBiometricLoginActivity::class.java))
+            }
+        }
+    }
+
+
+    // BIOMETRICS SECTION
+
+    private fun showBiometricPromptForDecryption() {
+        val canAuthenticate = BiometricManager.from(applicationContext).canAuthenticate()
+        if (canAuthenticate == BiometricManager.BIOMETRIC_SUCCESS) {
+            val secretKeyName = getString(R.string.secret_key_name)
+            val cipher = cryptographyManager.getInitializedCipherForDecryption(
+                secretKeyName, encryptedServerTokenWrapper.initializationVector
+            )
+            biometricPrompt =
+                BiometricPromptUtils.createBiometricPrompt(this, ::decryptServerTokenFromStorage)
+            val promptInfo = BiometricPromptUtils.createPromptInfo(this)
+            biometricPrompt.authenticate(promptInfo, BiometricPrompt.CryptoObject(cipher))
+        }
+    }
+
+    private fun decryptServerTokenFromStorage(authResult: BiometricPrompt.AuthenticationResult) {
+        authResult?.cryptoObject?.cipher?.apply {
+            val plaintext =
+                cryptographyManager.decryptData(encryptedServerTokenWrapper.ciphertext, this)
+            SampleAppUser.fakeToken = plaintext
+            // now that you have the token, you can query server for everything else
+            // the only reason we call this fakeToken is because we didn't really get it from
+            // the server. In your case, you will have gotten it from the server the first time
+            // and therefore, it's a real token.
+
+            updateApp(getString(R.string.already_signedin))
+        }
+    }
+
+    // USERNAME + PASSWORD SECTION
+
+    private fun loginWithPasswordManager() {
+        loginWithPasswordViewModel =
+            ViewModelProviders.of(this).get(LoginWithPasswordViewModel::class.java)
+        loginWithPasswordViewModel.loginWithPasswordFormState.observe(this, Observer {
+            val loginState = it ?: return@Observer
+            when(loginState){
+                is SuccessfulLoginFormState -> binding.login.isEnabled = loginState.isDataValid
+                is FailedLoginFormState ->{
+                    loginState.usernameError?.let { binding.username.error = getString(it) }
+                    loginState.passwordError?.let { binding.password.error = getString(it) }
+                }
+            }
+        })
+
+        binding.username.afterTextChanged {
+            loginWithPasswordViewModel.onLoginDataChanged(
+                binding.useBiometrics.text.toString(),
+                binding.password.text.toString()
+            )
+        }
+
+        binding.password.apply {
```

Comment:
`apply` isn't the [right scope operator to use](https://kotlinlang.org/docs/reference/scope-functions.html#function-selection) here (`apply` is only for configuring objects when they're created).

Either split them up into two statements with `binding.password.afterTextChanged`... so no scope function is required or use either `with` or `let`.

Reply:
As I mention elsewhere, the Android Studio Login Activity uses it this way. Clearly the fashion has changed since then, but `with` or `let` would not work here. So I can just split it.

Reply:
I agree with @nic0lette. While `apply` saves you one `binding.password` call it's not idiomatic. Please change.